### ISSUE TITLE
Static function call syntax for structs creates ambiguous code, without <Struct as Struct>

### ISSUE THREAD

**Issue by richardwhiuk** [rust-lang/rust#75698]
I tried this code:

```rust
struct Foo;
impl Foo {
    fn get(&self) {
        println!("!!");
    }

    fn foo_bar(&self) {
        println!("whoopsie");
    }
}
trait Bar {
    fn get(&self);
    fn foobar(&self);
}
impl Bar for Foo {

    fn get(&self) {
        Foo::get(self);
    }

    fn foobar(&self) {
        Foo::foobar(self);
    }
}

fn main() {
    
    let f = Foo;
    
    f.get();
    //f.foobar();
}
```

I expected to either see:
- An error from Rust, because `Foo::foobar` is not defined, and it would expect `Bar::foobar` in this scenario.
- An error from Rust, because `Foo::get()` is ambiguous.

Instead, this happened:

- `Bar::foobar` calls recursively.
- `Bar::get` calls `Foo::get`.

So the output is: `!!`. With the line commented out in, the code re-curses forever.

The only way to avoid this is to write both functions as: `<Foo as Foo>::get()`, and `<Foo as Foo>::foobar()`, at which point it does complain about `foobar` not being defined.

### Meta

Reproduces on the Playground, on Stable on 1.45.2 - https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3641b0fb4fd1e2c9702081ac0ddc44cb

**Comment by lzutao** [comment]
There is warnings when you compile the snippet:
```rust
warning: function cannot return without recursing
  --> src/main.rs:21:5
   |
21 |     fn foobar(&self) {
   |     ^^^^^^^^^^^^^^^^ cannot return without recursing
22 |         Foo::foobar(self);
   |         ----------------- recursive call site
   |
   = note: `#[warn(unconditional_recursion)]` on by default
   = help: a `loop` may express intention better if this is on purpose
```

**Comment by jonas-schievink** [comment]
> An error from Rust, because `Foo::foobar` is not defined, and it would expect `Bar::foobar` in this scenario.

It is defined through the `Bar` trait impl. It is intentional that `Foo::foobar` resolves to trait impls in scope when there is no inherent method defined. As such, closing as expected behavior.

**Comment by richardwhiuk** [comment]
Shouldn't the ambiguous call be an error?

**Comment by jonas-schievink** [comment]
Inherent impls take precedence over trait impls, so that also behaves as expected
